### DISCUSSION TITLE
Please allow 1.x and 2.x/3.x code to co-exist in the filesystem

### DISCUSSION THREAD

**Discussion by hartwork** [sqlalchemy/sqlalchemy#9750] [Usage Questions]
### Describe the use case

Hi!

I'm helping out with PyChess, a Desktop chess app at home at https://github.com/pychess/pychess. We recently migrated from SQLalchemy 1 to SQLalchemy 2 and that made us notice that some distributions cannot upgrade from PyChess 1.0.3 (using SQLalchemy 1) to PyChess 1.0.4 (using SQLalchemy 2) because upgrading to SQLalchemy 2 system-wide breaks other existing packages. Now if SQLalchemy 2 was made to use different filename and import paths, 1.x and 2.x could co-exist and upgrading would be painless to distributions; currently distros are forced to e.g. fix all applications at once or have two conflicting packages (e.g. see https://github.com/void-linux/void-packages/pull/43529) so that it will be impossible to use any application using 1.x while any using 2.x are installed. That's rather unfortunate.

To get out of that situation, one solution would be to move straight to 3.x in a way where all files can co-exist with 1.x, a bit like what projectM did for 3.x and 4.x (https://github.com/projectM-visualizer/projectm/pull/699). The move to 3.x seems to be needed to not break with current 2.x code.

Are you open for a direction like this?
Are there other ideas to ease the blocking situation that distros are facing today?
Are there any compatibility layers that I am not aware of?

Thanks in advance! :pray: 

### Databases / Backends / Drivers targeted

n/a

### Example Use

n/a

### Additional context

n/a

**Comment by zzzeek** [comment]
hi -

The design direction of SQLAlchemy 2, that we would break compat with SQLAlchemy 1.4 and also keep the same import namespace, was decided years ago, and this direction at the Python library level can't just be "changed".    If SQLAlchemy 2 used a new namespace, that would have been much worse for compatibility as you would now have environments using both SQLAlchemy versions simultaneously and it would be about five minutes before a few thousand users tried to mix Core and ORM constructs within the two libraries, which would fail.   So had we gone that way, we would instead be here fielding requests for "please make runtime objects cross-compatible between SQLAlchemy 1 and 2", which would be insurmountable enough that the whole SQLA 2 thing would just have to have not really happened.

now onto Pychess, where we will make the first observation which is that it's distributing itself as a system-wide distro, which is of course fine, but that introduces a lot of constraints on such a project.  As you likely know, the usual way Python projects with dependencies are installed is using virtualenvs, so that each environment has the specific versions of libraries that are needed.  but when you are targeting being a platform-wide dist that has to rely upon the other system wide Python libraries, you necessarily have to be very conservative about things, and at that level it seems like a version bump on the "patch" level, from 1.0.3 to 1.0.4, would not have been the appropriate place to do a major dependency change like this, at least not without full attention to backwards compatibility (and even then, still seems like a big change).     Beyond that, it's a given that system-wide libraries are going to change at different rates, so really the way this has to be approached for a system-wide dist is that your code has to dynamically adjust to either version present.    This is very doable with SQLAlchemy 2 as you can write apps that are fully compatible with SQLAlchemy 1.4 and 2.0 at the same time without even using any conditional logic.   Using conditionals you can do even more, things like using `class Base(DeclarativeBase):` with SQLAlchemy 2 and `Base = declarative_base()` with SQLAlchemy 1.x.

Basically I think you need to target 1.4 and 2.0 simultaneously not much differently than you'd target 1.3 and 1.4 simultaneously, if those were your target versions.  This is not really unusual or unreasonable.

**Reply by hartwork** [reply]
Hi @zzzeek,

thanks for your reply.  I only saw it now because the issue I created was converted into a  discussion where I apparently do not get e-mail notifications for replies out of the box…

The way I read your reply, there is no hope or help to distros other than mass-patching existing software or accepting mass conflicts regarding same-time installability of packages. That's a rather sad picture, and the reason why I started this issue/discussion.

What would be cool to understand is how using Alchemy 1.x and 2.x from the same process would have been a problem assuming fully disjoint filenames? How/why would runtime objects from both worlds need to be cross-compatible if it's two different databases?

Best, Sebastian

**Reply by CaselIT** [reply]
> The way I read your reply, there is no hope or help to distros other than mass-patching existing software or accepting mass conflicts regarding same-time installability of packages. That's a rather sad picture, and the reason why I started this issue/discussion.

Personally I think the issue here is not of sqlalchemy. If distro maintainers want to use a single shared env with all packages I guess it's their problem.

Also no package that I know do anything similar to what you are suggesting, so even if sqlalchemy were to do something like this I don't see how it would solve the issue.

Personally here the only problem I see is wanting to use a single version for every package instead of using different envs.

**Reply by zzzeek** [reply]
> 
> The way I read your reply, there is no hope or help to distros other than mass-patching existing software or accepting mass conflicts regarding same-time installability of packages. 

I dont see any reason why either of those things needs to be the case.    SQLAlchemy releases have been deprecating APIs and behaviors and removing them in place of new ones for 17 years.  There is nothing novel about this process nor has it ever implied "mass patching" or "mass conflicts", the changes are always fairly incremental, backwards compatibility and incremental upgrade paths are always well provided and documented, "hard" API and behavioral changes without any forwards compatibility pattern are kept to an absolute minimum, and the userbase and distrobase has navigated through every version without too many problems or any need to fork distributions, .. 0.4, 0.5, 0.6, 0.7 ...1.1, 1.2, 1.3, 1.4, 2.0.    There's nothing particularly different about SQLAlchemy 2.0 in this regard, all the API patterns that were removed were deprecated at least in version 1.4 and the vast majority of these patterns have been strongly discouraged and de-emphasized in the documentation for many years, such as "bound metadata" and "connectionless execution".  

More succinctly, as I said in my previous response, it is straightforward to have an application that runs on SQLAlchemy 1.4 and 2.0 at the same time.     There's a whole [migration guide](https://docs.sqlalchemy.org/en/20/changelog/migration_20.html) which documents how to achieve this.

What's not possible however is to write an application that uses SQLAlchemy 2.0's newest features, like ORM Annotated Declarative, and have that work with SQLAlchemy 1.4.   This is just the reality of having software that targets distributions.    We here in SQLAlchemy had to forego using terrific Python 3 features for many, many years after Python 3 was available and for which we supplied full support; because we still were providing 2.7 compatibility.    Right now, there's tons of Python features and APIs that have been available for many years, like positional-only function arguments, that we simply can't use until we can retire Python 3.7, 3.8 etc. support.    We will always be straddling the latest Python version and a many-years-old Python version at the same time.

If you look at Linux distros, you will see that they tend to move along with SQLAlchemy releases about 6 months later.  Current distros are all packaging SQLAlchemy 1.4 now, no longer 1.3, and within a year or two current distros will all be in 2.0.    There's code that runs on SQLAlchemy 1.4 that won't run on 1.3, there's 1.3 code that won't run on 1.2, etc.  


> 
> What would be cool to understand is how using Alchemy 1.x and 2.x from the same process would have been a problem assuming fully disjoint filenames? 

it firstly would make it impossible to write cross-compatible code for SQLAlchemy 1.x and 2.0, which is the norm.  This would stop adoption of 2.0 in its tracks, as it's essential that applications and libraries are able to target the previous and current SQLAlchemy series at the same time.    Openstack, which I work on, would have literally no upgrade path for 2.0 as it would require a truly "massive" change to tens of thousands of source files across dozens of separate but interrelated projects at the same time the library is upgraded.  The upgrade would be immensely difficult.

Beyond that, it immediately implies that you can have one program that imports from both modules, which could be that certain dependencies of the program use the newer or older version, etc..   The fact that these two modules have entirely different internal architectures would be immediately problematic, as we can't exactly say, "if your code uses models defined in "sqlalchemy2", we don't support them running with "sqlalchemy" constructs, the behavior will be undefined", people would be doing it immediately, if not directly, then as a result of dependencies making this happen, and they'd demand to know why it doesn't work, or that errors are fully specific and not undefined.

But overall the burden it would place on applications such as Openstack and libraries such as our own Alembic migrations to somehow support both packages simultaneously would be unreasonable and unnecessary.    There's no reason to make a separate import space, sqlalchemy 2.0 APIs can be used in a way that is backwards compatible with 1.4 in the same way that 1.4 has backwards compat with 1.3, 1.3 has compat with 1.2, etc.
